### PR TITLE
Feature/base list

### PR DESCRIPTION
**Pull request by MambetniyazovAmir**

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -26,6 +26,8 @@ dependencies {
     implementation 'androidx.appcompat:appcompat:1.0.2'
     implementation 'androidx.core:core-ktx:1.1.0'
     implementation 'androidx.constraintlayout:constraintlayout:1.1.3'
+    // recyclerView library
+    implementation 'com.android.support:recyclerview-v7:28.0.0'
```
Comment (atabekm) [inline app/build.gradle]: This is the old version of the lib, you need to use `androidx` version. And no need to comment above, it is clear from library name what it is.

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -3,15 +3,17 @@
...
 
     <application
-        android:allowBackup="true"
+        android:allowBackup="false"
         android:icon="@mipmap/ic_launcher"
         android:label="@string/app_name"
         android:roundIcon="@mipmap/ic_launcher_round"
         android:supportsRtl="true"
         android:theme="@style/AppTheme">
+        <activity android:name=".list.ListFragment" />
```
Comment (atabekm) [inline app/src/main/AndroidManifest.xml]: Only activities are required to be listed in manifest.

File: app/src/main/java/com/shagalalab/constitution/list/ListFragment.kt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import com.shagalalab.constitution.R
+import com.shagalalab.constitution.list.adapter.ItemClickListener
+import com.shagalalab.constitution.list.adapter.ListAdapter
+import com.shagalalab.constitution.list.model.Item
+import kotlinx.android.synthetic.main.fragment_list.*
+
+class ListFragment : AppCompatActivity(), ItemClickListener {
+
+    companion object {
+        const val k = 10
```
Comment (atabekm) [inline app/src/main/java/com/shagalalab/constitution/list/ListFragment.kt]: 1. Try to give meaningful names for variables instead of some random letters
2. Usually `const val` properties are always ALL_CAPS

File: app/src/main/java/com/shagalalab/constitution/list/ListFragment.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    private var lang = ""
+    private val adapter = ListAdapter(this)
```
Comment (atabekm) [inline app/src/main/java/com/shagalalab/constitution/list/ListFragment.kt]: 1. I don't see much value in this _global_ `adapter` variable, as it is only used inside `onCreate`, maybe you can have local variable inside `onCreate`.
2. Add space below. I will repeat again, there should be always empty line between methods and between variables and methods.

File: app/src/main/java/com/shagalalab/constitution/list/ListFragment.kt [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.fragment_list)
+        lang = intent.getStringExtra("lang")
+        list.adapter = adapter
+        list.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
```
Comment (atabekm) [inline app/src/main/java/com/shagalalab/constitution/list/ListFragment.kt]: It is also possible to set layout manager via xml layout, I'd suggest you to do that, as it decreases the lines here.

File: app/src/main/java/com/shagalalab/constitution/list/ListFragment.kt [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        adapter.setData(getData())
+    }
+
+    private fun getData(): MutableList<Item> {
+        val result: MutableList<Item> = arrayListOf()
+        result.add(Item("$lang-title-0", ""))
+        for (item in 1..k) {
+            result.add(Item("$lang-title-$item", "$lang-description-$item"))
+        }
+        return result
```
Comment (atabekm) [inline app/src/main/java/com/shagalalab/constitution/list/ListFragment.kt]: Here is a shorter version of the method:
```
return List(k) {
    Item("$lang-title-$it", if (Random.nextBoolean()) "$lang-description-$it" else "")
}
```
In general, kotlin is much concise language than java, most of the java expressions can be shortened in kotlin. After you've finished the task, try to look for places where you can do improvements.

File: app/src/main/java/com/shagalalab/constitution/list/ListFragment.kt [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.fragment_list)
+        lang = intent.getStringExtra("lang")
+        list.adapter = adapter
+        list.layoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
+        list.addItemDecoration(DividerItemDecoration(this, DividerItemDecoration.VERTICAL))
+        adapter.setData(getData())
+    }
+
+    private fun getData(): MutableList<Item> {
```
Comment (atabekm) [inline app/src/main/java/com/shagalalab/constitution/list/ListFragment.kt]: Do you know the difference between `List` and `MutableList`? Basically, `List` is immutable, that means once you create a list, you cannot change it. `MutableList` is as its name says mutable, you can change the list whatever way you'd like.

Here, you don't change the outcome of `getData()` method, you just pass it to adapter. That means, the signature of this method can be just `List`, and inside the method you can use `MutableList` if you wish. This will work, because `MutableList` is subclass (subinterface to be specific) of `List`.

File: app/src/main/java/com/shagalalab/constitution/list/adapter/ItemViewHolder.kt [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+import android.view.View
+import android.widget.TextView
+import androidx.recyclerview.widget.RecyclerView
+import com.shagalalab.constitution.R
+import com.shagalalab.constitution.list.model.Item
+
+class ItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+
+    var title: TextView = itemView.findViewById(R.id.title_text)
+    var description: TextView = itemView.findViewById(R.id.description_text)
```
Comment (atabekm) [inline app/src/main/java/com/shagalalab/constitution/list/adapter/ItemViewHolder.kt]: You don't need to use `findViewById` in kotlin, remove them, and use the followings in `populateModel`:
- `itemView.title_text`
- `itemView.description_text`

File: app/src/main/java/com/shagalalab/constitution/list/adapter/ItemViewHolder.kt [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+    fun populateModel(model: Item, itemClickListener: ItemClickListener) {
+        title.text = model.title
+        description.text = model.description
+        if (model.description == "") {
+            description.visibility = View.GONE
```
Comment (atabekm) [inline app/src/main/java/com/shagalalab/constitution/list/adapter/ItemViewHolder.kt]: There is a `isVisible` extension function for `View` to which you pass boolean statement, then there is no need for `if` condition below.

File: app/src/main/java/com/shagalalab/constitution/list/adapter/ItemViewHolder.kt [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+class ItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+
+    var title: TextView = itemView.findViewById(R.id.title_text)
+    var description: TextView = itemView.findViewById(R.id.description_text)
+
+    fun populateModel(model: Item, itemClickListener: ItemClickListener) {
+        title.text = model.title
+        description.text = model.description
+        if (model.description == "") {
```
Comment (atabekm) [inline app/src/main/java/com/shagalalab/constitution/list/adapter/ItemViewHolder.kt]: Do you know that there is `isEmpty()` extension function for String class, use it instead.

File: app/src/main/res/layout/fragment_list.xml [hunk 11]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (atabekm) [inline app/src/main/res/layout/fragment_list.xml]: Why do you need this `ConstraintLayout` if it has only one child?

File: app/src/main/res/layout/fragment_list_item.xml [hunk 12]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="64dp">
+
+    <LinearLayout
```
Comment (atabekm) [inline app/src/main/res/layout/fragment_list_item.xml]: Why do you need this `LinearLayout` if it has `ConstraintLayout` as a parent? Use one of them.

File: app/src/main/res/layout/fragment_list_item.xml [hunk 13]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:layout_width="match_parent"
+    android:layout_height="64dp">
```
Comment (atabekm) [inline app/src/main/res/layout/fragment_list_item.xml]: Extract hardcoded dimension to dimens.xml

File: app/src/main/res/values/colors.xml [hunk 14]
```diff
@@ -4,4 +4,6 @@
     <color name="colorPrimaryDark">#00574B</color>
     <color name="colorAccent">#D81B60</color>
     <color name="blue">#3A63D3</color>
+    <color name="black">#000000</color>
+    <color name="blackish">#2A2A2A</color>
```
Comment (atabekm) [inline app/src/main/res/values/colors.xml]: Bad naming, if you don't know the color name, then it is better to name it according to its used place, e.g. `main_text_color` or similar.

File: app/src/main/res/layout/fragment_list_item.xml [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+            android:textSize="@dimen/title_text_size" />
+
+        <TextView
+            android:id="@+id/description_text"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginStart="@dimen/standard_margin"
+            android:layout_marginEnd="@dimen/standard_margin"
+            android:text="@string/description"
+            android:textColor="@color/blackish"
```
Comment (atabekm) [inline app/src/main/res/layout/fragment_list_item.xml]: This color is too dark, I don't see any difference between the above one. Change the color to lighter one.

File: app/src/main/res/values/dimens.xml [hunk 16]
```diff
@@ -3,5 +3,6 @@
     <dimen name="flag_size">110dp</dimen>
     <dimen name="standard_margin">16dp</dimen>
     <dimen name="big_margin">24dp</dimen>
-    <dimen name="text_size">20sp</dimen>
+    <dimen name="text_size">16sp</dimen>
+    <dimen name="title_text_size">20sp</dimen>
```
Comment (atabekm) [inline app/src/main/res/values/dimens.xml]: Too big, `18` seems better.

File: app/src/main/res/layout/fragment_list_item.xml [hunk 17]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (atabekm) [inline app/src/main/res/layout/fragment_list_item.xml]: Wrong file naming, this layout is not directly related to fragment, it is the layout for adapter. The convention is to name it `item_*` for ViewHolder layouts.

**Review by atabekm — CHANGES_REQUESTED** [review]

File: app/src/main/AndroidManifest.xml (hunk 2 above)
Comment (MambetniyazovAmir) [reply app/src/main/AndroidManifest.xml]: I mistakenly made an activity, now redid it to a fragment, and if I am not it should be only fragments and only one main activity am I right?

Comment (MambetniyazovAmir) [reply app/src/main/AndroidManifest.xml]: But there is UI problem when opening listFragment it just changes mainfragment with listFragment it seems bad, activity was more pleasant.

File: app/src/main/res/layout/fragment_list_item.xml (hunk 12 above)
Comment (MambetniyazovAmir) [reply app/src/main/res/layout/fragment_list_item.xml]: I could not do it using only "ConstraintLayout" or only "LinearLayout", the problem is that when there is no "description" I had to have "title" in the middle, and when there was, they should be exactly in the middle. In what I did, they divide the free space on a level. Are any crashs possible in this case?

File: app/src/main/AndroidManifest.xml (hunk 2 above)
Comment (atabekm) [reply app/src/main/AndroidManifest.xml]: We are using single activity principle in this app, there will be MainActivity, and all screens are in fragments. Regarding no proper transition/animation between fragments, that can be done, we will do it later.

File: app/src/main/res/layout/fragment_list_item.xml (hunk 12 above)
Comment (atabekm) [reply app/src/main/res/layout/fragment_list_item.xml]: No, the problem is not with possible crash. It is just inefficient usage of layouts, basically both `LinearLayout` and `ConstraintLayout` are viewgroups (which holds other views), there is not much value on nesting them. `ConstraintLayout` is much more flexible, you should be able to achieve the behavior you want. Try to look at a few `ConstraintLayout` tutorials, or stackoverflow.

File: app/src/main/AndroidManifest.xml [hunk 18]
```diff
@@ -12,6 +12,7 @@
         <activity android:name=".MainActivity">
             <intent-filter>
                 <action android:name="android.intent.action.MAIN" />
+                <action android:name="android.intent.action.VIEW" />
```
Comment (atabekm) [inline app/src/main/AndroidManifest.xml]: I don't think we need this

File: build.gradle [hunk 19]
```diff
@@ -1,7 +1,7 @@
 // Top-level build file where you can add configuration options common to all sub-projects/modules.
 
 buildscript {
-    ext.kotlin_version = '1.3.50'
+    ext.kotlin_version = '1.3.41'
```
Comment (atabekm) [inline build.gradle]: Why you have downgraded this?

File: app/src/main/java/com/shagalalab/constitution/list/adapter/ItemViewHolder.kt [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+import kotlinx.android.synthetic.main.item_view.view.*
+
+class ItemViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+
+    var title: TextView = itemView.title_text
+    var description: TextView = itemView.description_text
+
+    fun populateModel(model: Item, itemClickListener: ItemClickListener) {
+        title.text = model.title
+        description.text = model.description
```
Comment (atabekm) [inline app/src/main/java/com/shagalalab/constitution/list/adapter/ItemViewHolder.kt]: you don't need a variable for `title` and `description`, you should instead directly call the view like `itemView.title_text.text = ...` and etc.

File: app/src/main/res/layout/fragment_list.xml [hunk 21]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="utf-8"?>
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (atabekm) [inline app/src/main/res/layout/fragment_list.xml]: You don't need this `LinearLayout`, just `RecyclerView` itself should be fine.

**Review by atabekm — CHANGES_REQUESTED** [review]
Hopefully this is the last review for this PR, if you resolve all, it should be ready for merge.

File: build.gradle (hunk 19 above)
Comment (MambetniyazovAmir) [reply build.gradle]: There was this "Kotlin version is used for building with Gradle (1.3.50) differs from the one bundled into the IDE plugin (1.3.41)"

Comment (atabekm) [reply build.gradle]: Then that means your kotlin version is outdated, do not downgrade it, instead update yourself

File: app/src/main/java/com/shagalalab/constitution/list/adapter/ListAdapter.kt [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+import android.view.LayoutInflater
+import android.view.ViewGroup
+import androidx.recyclerview.widget.RecyclerView
+import com.shagalalab.constitution.R
+import com.shagalalab.constitution.list.model.Item
+
+class ListAdapter(private val itemClickListener: ItemClickListener) :
+    RecyclerView.Adapter<ItemViewHolder>() {
+
+    var models: List<Item> = arrayListOf()
```
Comment (atabekm) [inline app/src/main/java/com/shagalalab/constitution/list/adapter/ListAdapter.kt:12]: I will merge for now, but pay attention to the warnings from Android studio, here for example it is saying this can become `private`. You can fix it with your next task.

File: app/src/main/res/layout/fragment_list.xml [hunk 23]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.recyclerview.widget.RecyclerView xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:app="http://schemas.android.com/apk/res-auto"
+    android:id="@+id/list"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:background="@color/white"
+    android:orientation="vertical"
```
Comment (atabekm) [inline app/src/main/res/layout/fragment_list.xml:8]: This is LinearLayout's parameter, you don't need it here. Please remove with your next task.

**Review by atabekm — APPROVED** [review]
Two minor things, but they are not a blocker, so I will merge, but fix them with your next task.
